bpf/lib/encrypt.h

```diff
@@ -204,6 +205,120 @@ do_decrypt(struct __ctx_buff *ctx, __u16 proto)
 #endif /* ENABLE_ENDPOINT_ROUTES */
 }
 
+/* checks whether a IPsec redirect should be performed for the security id
+ * we do not IPsec encrypt:
+ * 1. Host-to-Host or Pod-to-Host traffic
+ * 2. Traffic leaving the cluster
+ * 3. Remote nodes including Kube API server
+ */
+static __always_inline int
+ipsec_redirect_sec_id_ok(__u32 sec_id) {
+	if (sec_id == HOST_ID)
+		return 0;
+	if (!identity_is_cluster(sec_id))
+		return 0;
+	if (identity_is_remote_node(sec_id))
+		return 0;
+	return 1;
+}
+
+static __always_inline int
+ipsec_maybe_redirect_to_encrypt(struct __ctx_buff *ctx, __be16 proto)
+{
+	struct remote_endpoint_info __maybe_unused *dst = NULL;
+	struct remote_endpoint_info __maybe_unused *src = NULL;
+	void *data __maybe_unused, *data_end __maybe_unused;
+	struct iphdr __maybe_unused *ip4;
+	struct ipv6hdr __maybe_unused *ip6;
+	__u32 magic __maybe_unused = 0;
+	int ret = 0;
+	union macaddr dst_mac = CILIUM_NET_MAC;
+
+	if (!eth_is_supported_ethertype(proto))
+		return DROP_UNSUPPORTED_L2;
+
+	switch (proto) {
+# ifdef ENABLE_IPV4
+	case bpf_htons(ETH_P_IP):
+		if (!revalidate_data(ctx, &data, &data_end, &ip4))
+			return DROP_INVALID;
+
+		dst = lookup_ip4_remote_endpoint(ip4->daddr, 0);
+		src = lookup_ip4_remote_endpoint(ip4->saddr, 0);
+
+#  if defined(TUNNEL_MODE)
+		/* tunnel mode needs a bit of special handling when
+		 * encapsulated packets get here the destination address is
+		 * already a cluster node IP.
+		 *
+		 * cluster node IPs will have 0.0.0.0 as their tunnel_endpoint
```

Comment:
Could you double-check this part? I believe it changed with https://github.com/cilium/cilium/pull/36978.

Reply:
Yup, you are correct:

```
cilium bpf ipcache list | grep -E "^172.*"
172.18.0.2/32               identity=1 encryptkey=0 tunnelendpoint=0.0.0.0 flags=skiptunnel
172.18.0.3/32               identity=7 encryptkey=0 tunnelendpoint=172.18.0.3 flags=skiptunnel
172.18.0.4/32               identity=6 encryptkey=0 tunnelendpoint=172.18.0.4 flags=skiptunne
```

Let me see if this simplifies things...

Reply:
So, it does not clean up too much, since we still need to get the source sec ID from `ctx_is_overlay(ctx)` and subsequent `get_identity(ctx)`. 

I think at the most we can update to:

```
		if (ctx_is_overlay(ctx)) {
			ret = set_ipsec_encrypt(ctx, 0, dst->tunnel_endpoint,
						get_identity(ctx), false,
						true);
			if (ret != CTX_ACT_OK)
				return ret;
			goto overlay_encrypt;
		}
```

But not sure this gains us that much. 

At the least, I can update the comments.

Reply:
Hmm, actually testing the above, we run into some errors. 🤔

Reply:
Talked with @julianwiedmann out of band. Because we still need to get CTX from where overlay prog stuffs it, we will simply update the comments here, as they are incorrect and misleading.